Source/UI/Hashtag/HashtagView.swift

```diff
@@ -0,0 +1,145 @@
+import SwiftUI
+
+@MainActor
+struct HashtagView: View {
+    
+    var hashtag: Hashtag
+
+    private var strategy: FeedStrategy {
+        HashtagAlgorithm(hashtag: hashtag)
+    }
+
+    @State
+    var messages: [Message]?
+
+    @State
+    private var isLoadingMoreMessages = false
+
+    @State
+    private var offset = 0
+
+    @State
+    private var noMoreMessages = false
+
+    @EnvironmentObject
+    private var botRepository: BotRepository
+
+    var body: some View {
+        Group {
+            if let messages = messages {
+                ScrollView(.vertical, showsIndicators: false) {
+                    ZStack {
+                        LazyVStack {
+                            if messages.isEmpty {
+                                EmptyPostsView(description: Localized.Message.noPostsInHashtagDescription)
+                            } else {
+                                ForEach(messages) { message in
+                                    Button {
+                                        if let contact = message.content.contact {
+                                            AppController.shared.open(identity: contact.contact)
+                                        } else {
+                                            AppController.shared.open(identifier: message.id)
+                                        }
+                                    } label: {
+                                        MessageView(message: message)
+                                            .onAppear {
+                                                if message == messages.last {
+                                                    loadMore()
+                                                }
+                                            }
+                                    }
+                                    .buttonStyle(MessageButtonStyle())
+                                }
+                            }
+                        }
+                        .frame(maxWidth: 500)
+                        .padding(EdgeInsets(top: 0, leading: 0, bottom: 15, trailing: 0))
+                        if isLoadingMoreMessages, !noMoreMessages {
+                            HStack {
+                                ProgressView().frame(maxWidth: .infinity, alignment: .center).padding()
+                            }
+                        }
+                    }
+                    .frame(maxWidth: .infinity)
+                }
+            } else {
+                LoadingView()
+            }
+        }
+        .task {
+            await loadFromScratch()
+        }
+        .refreshable {
+            await loadFromScratch()
+        }
+        .background(Color.appBg)
+        .navigationTitle(hashtag.string)
+    }
+
+    func loadFromScratch() async {
+        let bot = botRepository.current
+        let pageSize = 50
+        do {
+            let newMessages = try await bot.feed(strategy: strategy, limit: pageSize, offset: 0)
```

Comment:
It looks like a lot of code is duplicated between `loadFromScratch()` and `loadMore()`. This pattern is repeated in other files as well. Could you refactor these to keep the code DRY?

I think we could even create a component called `InfiniteList` or something that would extract out a lot of the common functionality in our message list views. Showing and hiding the loading indicator, showing the empty message, loading more rows when we get to the bottom, etc. I think I had an `InfiniteList` view started on my SwiftUI branch I had started on before paternity leave. This is larger task though. Maybe best to open a new ticket for it?

Reply:
This is being done in a future PR